compiler/rustc_lint/src/non_fmt_panic.rs

```diff
@@ -0,0 +1,197 @@
+use crate::{LateContext, LateLintPass, LintContext};
+use rustc_ast as ast;
+use rustc_errors::{pluralize, Applicability};
+use rustc_hir as hir;
+use rustc_middle::ty;
+use rustc_parse_format::{ParseMode, Parser, Piece};
+use rustc_span::{sym, symbol::kw, InnerSpan, Span, Symbol};
+
+declare_lint! {
+    /// The `non_fmt_panic` lint detects `panic!(..)` invocations where the first
+    /// argument is not a formatting string.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,no_run
+    /// panic!("{}");
+    /// panic!(123);
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// In Rust 2018 and earlier, `panic!(x)` directly uses `x` as the message.
+    /// That means that `panic!("{}")` panics with the message `"{}"` instead
+    /// of using it as a formatting string, and `panic!(123)` will panic with
+    /// an `i32` as message.
+    ///
+    /// Rust 2021 always interprets the first argument as format string.
+    NON_FMT_PANIC,
```

Comment:
I'm slightly concerned: we are changing the name of an existing lint. What will happen to anyone that has already used `#![deny(fmt_panic)]` or `#![allow(fmt_panic)]`?

Reply:
That will show ``warning: unknown lint: `fmt_panic` ``. `fmt_panic` has only been there since .1.50, which is still in beta. I suppose we could also change the name there before it gets to stable. But I don't know how important this is. The lint does new things, so using the same name might also not be a great idea, as `deny` or `allow` will then deny or allow more things than before.

Reply:
https://github.com/rust-lang/rust/pull/81729